src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,367 @@
+  rw [hb, jacobi_sym_zero_right],
+  { exact one_ne_zero },
+  { contrapose! h, exact jacobi_sym_eq_zero_iff_not_coprime.1 h },
+end
+
+/-- The Jacobi symbol `(a / b)` vanishes if and only if `b ≠ 0` and `a` and `b` are not coprime. -/
+lemma jacobi_sym_eq_zero_iff {a : ℤ} {b : ℕ} : [a | b]ⱼ = 0 ↔ b ≠ 0 ∧ a.gcd b ≠ 1 :=
+⟨λ h, begin
+  casesI eq_or_ne b 0 with hb hb,
+  { rw [hb, jacobi_sym_zero_right] at h, cases h },
+  exact ⟨hb, mt jacobi_sym_ne_zero $ not_not.2 h⟩,
+end, λ ⟨hb, h⟩, by { rw ← ne_zero_iff at hb, exactI jacobi_sym_eq_zero_iff_not_coprime.2 h }⟩
+
+/-- The Jacobi symbol `(0 / b)` vanishes when `b > 1`. -/
+lemma jacobi_sym_zero_left {b : ℕ} (hb : 1 < b) : [0 | b]ⱼ = 0 :=
+(@jacobi_sym_eq_zero_iff_not_coprime 0 b ⟨ne_zero_of_lt hb⟩).mpr $
+  by { rw [int.gcd_zero_left, int.nat_abs_of_nat], exact hb.ne' }
+
+/-- The Jacobi symbol `(a / b)` takes the value `1` or `-1` if `a` and `b` are coprime. -/
+lemma jacobi_sym_eq_one_or_neg_one {a : ℤ} {b : ℕ} (h : a.gcd b = 1) :
+  [a | b]ⱼ = 1 ∨ [a | b]ⱼ = -1 :=
+(jacobi_sym_trichotomy a b).resolve_left $ jacobi_sym_ne_zero h
+
+/-- We have that `(a^e / b) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_left (a : ℤ) (e b : ℕ) : [a ^ e | b]ⱼ = [a | b]ⱼ ^ e :=
+nat.rec_on e (by rw [pow_zero, pow_zero, jacobi_sym_one_left]) $
+  λ _ ih, by rw [pow_succ, pow_succ, jacobi_sym_mul_left, ih]
+
+/-- We have that `(a / b^e) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_right (a : ℤ) (b e : ℕ) : [a | b ^ e]ⱼ = [a | b]ⱼ ^ e :=
+begin
+  induction e with e ih,
+  { rw [pow_zero, pow_zero, jacobi_sym_one_right], },
+  { casesI eq_zero_or_ne_zero b with hb,
+    { rw [hb, zero_pow (succ_pos e), jacobi_sym_zero_right, one_pow], },
+    { rw [pow_succ, pow_succ, jacobi_sym_mul_right, ih], } }
+end
+
+/-- The square of the Jacobi symbol `(a / b)` is `1` when `a` and `b` are coprime. -/
+lemma jacobi_sym_sq_one {a : ℤ} {b : ℕ} (h : a.gcd b = 1) : [a | b]ⱼ ^ 2 = 1 :=
+by cases jacobi_sym_eq_one_or_neg_one h with h₁ h₁; rw h₁; refl
+
+/-- The Jacobi symbol `(a^2 / b)` is `1` when `a` and `b` are coprime. -/
+lemma jacobi_sym_sq_one' {a : ℤ} {b : ℕ} (h : a.gcd b = 1) : [a ^ 2 | b]ⱼ = 1 :=
+by rw [jacobi_sym_pow_left, jacobi_sym_sq_one h]
+
+/-- The Jacobi symbol `(a / b)` depends only on `a` mod `b`. -/
+lemma jacobi_sym_mod_left (a : ℤ) (b : ℕ) : [a | b]ⱼ = [a % b | b]ⱼ :=
+congr_arg list.prod $ list.pmap_congr _ begin
+  rintro p hp _ _,
+  conv_rhs { rw [legendre_sym_mod, int.mod_mod_of_dvd _
+    (int.coe_nat_dvd.2 $ dvd_of_mem_factors hp), ← legendre_sym_mod] },
+end
+
+/-- The Jacobi symbol `(a / b)` depends only on `a` mod `b`. -/
+lemma jacobi_sym_mod_left' {a₁ a₂ : ℤ} {b : ℕ} (h : a₁ % b = a₂ % b) : [a₁ | b]ⱼ = [a₂ | b]ⱼ :=
+by rw [jacobi_sym_mod_left, h, ← jacobi_sym_mod_left]
+
+/-- If the Jacobi symbol `(a / b)` is `-1`, then `a` is not a square modulo `b`. -/
+lemma nonsquare_of_jacobi_sym_eq_neg_one {a : ℤ} {b : ℕ} (h : [a | b]ⱼ = -1) :
```

Comment:
I think mathlib convention would prefer
```suggestion
lemma not_is_square_of_jacobi_sym_eq_neg_one {a : ℤ} {b : ℕ} (h : [a | b]ⱼ = -1) :
```
but I see that `nonsquare` has precedent in [finite_field.exists_nonsquare](https://leanprover-community.github.io/mathlib_docs/field_theory/finite/basic.html#finite_field.exists_nonsquare) so maybe it's OK.

Reply:
I think I'll leave it unless other people want to have it changed, too.